// jsIF97/IF97_BK2.mjs
/*******************************************************************/
/*  Tph2(P,h,T)                                                    */
/*  Tps2(P,s,T)                                                    */
/*  Tph2A(P,h,T)                                                   */
/*  Tph2B(P,h,T)                                                   */
/*  Tph2C(P,h,T)                                                   */
/*  Tps2A(P,h,T)                                                   */
/*  Tps2B(P,h,T)                                                   */
/*  Tps2C(P,h,T)                                                   */
/*******************************************************************/

/* backward function of region 2 */
/* based on Eqs.(22), (23), and (24) of IAPWS-IF97 */
"use strict"
export function Tph2(SP){
  var p585;
  
  if(SP.P <= 4.0){
    if(Tph2A(SP)==-1){SP = null;return -1;}
  }
  else{
    p585= (0.12809002730136e-3 * SP.h - 0.67955786399241e+0) * SP.h
         +0.90584278514723e+3;

    if (SP.P <= p585){
      if(Tph2B(SP)==-1){SP=null;return -1;}
    }
    else{
      if(Tph2C(SP)==-1){SP=null;return -1;}
    }
  }
  
  return 1;
}

export function Tps2(SP){
  
  if(SP.P <= 4.0){
    if(Tps2A(SP)==-1){SP = null;return -1;}
  }
  else{
    if (SP.s >= 5.85){
    if(Tps2B(SP)==-1){SP = null;return -1;}
    }
    else{    
    if(Tps2C(SP)==-1){SP = null;return -1;}
    }
  }  
  return 1;
}

function Tph2A(SP){
  var i;
  var eta;
  var Hn;
  var T;  
  var II = [];
  var JJ = [];
  var an = [];
  
  Hn= 2000.0;
  II[ 1]=  0  ;  JJ[ 1]=   0  ;  an[ 1]=   0.10898952318288e+4  ;
  II[ 2]=  0  ;  JJ[ 2]=   1  ;  an[ 2]=   0.84951654495535e+3  ;
  II[ 3]=  0  ;  JJ[ 3]=   2  ;  an[ 3]=  -0.10781748091826e+3  ;
  II[ 4]=  0  ;  JJ[ 4]=   3  ;  an[ 4]=   0.33153654801263e+2  ;
  II[ 5]=  0  ;  JJ[ 5]=   7  ;  an[ 5]=  -0.74232016790248e+1  ;
  II[ 6]=  0  ;  JJ[ 6]=  20  ;  an[ 6]=   0.11765048724356e+2  ;
  II[ 7]=  1  ;  JJ[ 7]=   0  ;  an[ 7]=   0.18445749355790e+1  ;
  II[ 8]=  1  ;  JJ[ 8]=   1  ;  an[ 8]=  -0.41792700549624e+1  ;
  II[ 9]=  1  ;  JJ[ 9]=   2  ;  an[ 9]=   0.62478196935812e+1  ;
  II[10]=  1  ;  JJ[10]=   3  ;  an[10]=  -0.17344563108114e+2  ;
  II[11]=  1  ;  JJ[11]=   7  ;  an[11]=  -0.20058176862096e+3  ;
  II[12]=  1  ;  JJ[12]=   9  ;  an[12]=   0.27196065473796e+3  ;
  II[13]=  1  ;  JJ[13]=  11  ;  an[13]=  -0.45511318285818e+3  ;
  II[14]=  1  ;  JJ[14]=  18  ;  an[14]=   0.30919688604755e+4  ;
  II[15]=  1  ;  JJ[15]=  44  ;  an[15]=   0.25226640357872e+6  ;
  II[16]=  2  ;  JJ[16]=   0  ;  an[16]=  -0.61707422868339e-2  ;
  II[17]=  2  ;  JJ[17]=   2  ;  an[17]=  -0.31078046629583e+0  ;
  II[18]=  2  ;  JJ[18]=   7  ;  an[18]=   0.11670873077107e+2  ;
  II[19]=  2  ;  JJ[19]=  36  ;  an[19]=   0.12812798404046e+9  ;
  II[20]=  2  ;  JJ[20]=  38  ;  an[20]=  -0.98554909623276e+9  ;
  II[21]=  2  ;  JJ[21]=  40  ;  an[21]=   0.28224546973002e+10 ;
  II[22]=  2  ;  JJ[22]=  42  ;  an[22]=  -0.35948971410703e+10 ;
  II[23]=  2  ;  JJ[23]=  44  ;  an[23]=   0.17227349913197E10  ;
  II[24]=  3  ;  JJ[24]=  24  ;  an[24]=  -0.13551334240775e+5  ;
  II[25]=  3  ;  JJ[25]=  44  ;  an[25]=   0.12848734664650e+8  ;
  II[26]=  4  ;  JJ[26]=  12  ;  an[26]=   0.13865724283226e+1  ;
  II[27]=  4  ;  JJ[27]=  32  ;  an[27]=   0.23598832556514e+6  ;
  II[28]=  4  ;  JJ[28]=  44  ;  an[28]=  -0.13105236545054e+8  ;
  II[29]=  5  ;  JJ[29]=  32  ;  an[29]=   0.73999835474766e+4  ;
  II[30]=  5  ;  JJ[30]=  36  ;  an[30]=  -0.55196697030060e+6  ;
  II[31]=  5  ;  JJ[31]=  42  ;  an[31]=   0.37154085996233e+7  ;
  II[32]=  6  ;  JJ[32]=  34  ;  an[32]=   0.19127729239660e+5  ;
  II[33]=  6  ;  JJ[33]=  44  ;  an[33]=  -0.41535164835634e+6  ;
  II[34]=  7  ;  JJ[34]=  28  ;  an[34]=  -0.62459855192507e+2  ;
  
  eta = SP.h / Hn;
  T = 0.0;
  for(i=1;i<=34;i++){
    T = T + an[i]*Math.pow(SP.P,II[i])*Math.pow(eta-2.1,JJ[i]);
  }
  SP.T=T;
  
  return 1;
}

function Tph2B(SP){
  var i;
  var eta;
  var Hn;
  var T;  
  var II = [];
  var JJ = [];
  var an = [];
  
  Hn= 2000.0;
  II[ 1]=  0  ;  JJ[ 1]=   0   ;  an[ 1]=   0.14895041079516e+4  ;
  II[ 2]=  0  ;  JJ[ 2]=   1   ;  an[ 2]=   0.74307798314034e+3  ;
  II[ 3]=  0  ;  JJ[ 3]=   2   ;  an[ 3]=  -0.97708318797837e+2  ;
  II[ 4]=  0  ;  JJ[ 4]=  12   ;  an[ 4]=   0.24742464705674e+1  ;
  II[ 5]=  0  ;  JJ[ 5]=  18   ;  an[ 5]=  -0.63281320016026e+0  ;
  II[ 6]=  0  ;  JJ[ 6]=  24   ;  an[ 6]=   0.11385952129658e+1  ;
  II[ 7]=  0  ;  JJ[ 7]=  28   ;  an[ 7]=  -0.47811863648625e+0  ;
  II[ 8]=  0  ;  JJ[ 8]=  40   ;  an[ 8]=   0.85208123431544e-2  ;
  II[ 9]=  1  ;  JJ[ 9]=   0   ;  an[ 9]=   0.93747147377932e+0  ;
  II[10]=  1  ;  JJ[10]=   2   ;  an[10]=   0.33593118604916e+1  ;
  II[11]=  1  ;  JJ[11]=   6   ;  an[11]=   0.33809355601454e+1  ;
  II[12]=  1  ;  JJ[12]=  12   ;  an[12]=   0.16844539671904e+0  ;
  II[13]=  1  ;  JJ[13]=  18   ;  an[13]=   0.73875745236695e+0  ;
  II[14]=  1  ;  JJ[14]=  24   ;  an[14]=  -0.47128737436186e+0  ;
  II[15]=  1  ;  JJ[15]=  28   ;  an[15]=   0.15020273139707e+0  ;
  II[16]=  1  ;  JJ[16]=  40   ;  an[16]=  -0.21764114219750e-2  ;
  II[17]=  2  ;  JJ[17]=   2   ;  an[17]=  -0.21810755324761e-1  ;
  II[18]=  2  ;  JJ[18]=   8   ;  an[18]=  -0.10829784403677e+0  ;
  II[19]=  2  ;  JJ[19]=  18   ;  an[19]=  -0.46333324635812e-1  ;
  II[20]=  2  ;  JJ[20]=  40   ;  an[20]=   0.71280351959551e-4  ;
  II[21]=  3  ;  JJ[21]=   1   ;  an[21]=   0.11032831789999e-3  ;
  II[22]=  3  ;  JJ[22]=   2   ;  an[22]=   0.18955248387902e-3  ;
  II[23]=  3  ;  JJ[23]=  12   ;  an[23]=   0.30891541160537e-2  ;
  II[24]=  3  ;  JJ[24]=  24   ;  an[24]=   0.13555504554949e-2  ;
  II[25]=  4  ;  JJ[25]=   2   ;  an[25]=   0.28640237477456e-6  ;
  II[26]=  4  ;  JJ[26]=  12   ;  an[26]=  -0.10779857357512e-4  ;
  II[27]=  4  ;  JJ[27]=  18   ;  an[27]=  -0.76462712454814e-4  ;
  II[28]=  4  ;  JJ[28]=  24   ;  an[28]=   0.14052392818316e-4  ;
  II[29]=  4  ;  JJ[29]=  28   ;  an[29]=  -0.31083814331434e-4  ;
  II[30]=  4  ;  JJ[30]=  40   ;  an[30]=  -0.10302738212103e-5  ;
  II[31]=  5  ;  JJ[31]=  18   ;  an[31]=   0.28217281635040e-6  ;
  II[32]=  5  ;  JJ[32]=  24   ;  an[32]=   0.12704902271945e-5  ;
  II[33]=  5  ;  JJ[33]=  40   ;  an[33]=   0.73803353468292e-7  ;
  II[34]=  6  ;  JJ[34]=  28   ;  an[34]=  -0.11030139238909e-7  ;
  II[35]=  7  ;  JJ[35]=   2   ;  an[35]=  -0.81456365207833e-13 ;
  II[36]=  7  ;  JJ[36]=  28   ;  an[36]=  -0.25180545682962e-10 ;
  II[37]=  9  ;  JJ[37]=   1   ;  an[37]=  -0.17565233969407e-17 ;
  II[38]=  9  ;  JJ[38]=  40   ;  an[38]=   0.86934156344163e-14 ;  
  eta = SP.h / Hn;
  T = 0.0;
  for(i=1;i<=38;i++){
    T = T + an[i]*Math.pow(SP.P-2.0,II[i])*Math.pow(eta-2.6,JJ[i]);
  }
  SP.T=T;
  
  return 1;
}

function Tph2C(SP){
  var i;
  var eta;
  var Hn;
  var T;  
  var II = [];
  var JJ = [];
  var an = [];
  
  Hn= 2000.0;
  II[ 1]=  -7   ;  JJ[ 1]=   0  ;  an[ 1]=  -0.32368398555242e+13  ;
  II[ 2]=  -7   ;  JJ[ 2]=   4  ;  an[ 2]=   0.73263350902181e+13  ;
  II[ 3]=  -6   ;  JJ[ 3]=   0  ;  an[ 3]=   0.35825089945447e+12  ;
  II[ 4]=  -6   ;  JJ[ 4]=   2  ;  an[ 4]=  -0.58340131851590e+12  ;
  II[ 5]=  -5   ;  JJ[ 5]=   0  ;  an[ 5]=  -0.10783068217470e+11  ;
  II[ 6]=  -5   ;  JJ[ 6]=   2  ;  an[ 6]=   0.20825544563171e+11  ;
  II[ 7]=  -2   ;  JJ[ 7]=   0  ;  an[ 7]=   0.61074783564516e+6   ;
  II[ 8]=  -2   ;  JJ[ 8]=   1  ;  an[ 8]=   0.85977722535580e+6   ;
  II[ 9]=  -1   ;  JJ[ 9]=   0  ;  an[ 9]=  -0.25745723604170e+5   ;
  II[10]=  -1   ;  JJ[10]=   2  ;  an[10]=   0.31081088422714e+5   ;
  II[11]=   0   ;  JJ[11]=   0  ;  an[11]=   0.12082315865936e+4   ;
  II[12]=   0   ;  JJ[12]=   1  ;  an[12]=   0.48219755109255e+3   ;
  II[13]=   1   ;  JJ[13]=   4  ;  an[13]=   0.37966001272486e+1   ;
  II[14]=   1   ;  JJ[14]=   8  ;  an[14]=  -0.10842984880077e+2   ;
  II[15]=   2   ;  JJ[15]=   4  ;  an[15]=  -0.45364172676660e-1   ;
  II[16]=   6   ;  JJ[16]=   0  ;  an[16]=   0.14559115658698e-12  ;
  II[17]=   6   ;  JJ[17]=   1  ;  an[17]=   0.11261597407230e-11  ;
  II[18]=   6   ;  JJ[18]=   4  ;  an[18]=  -0.17804982240686e-10  ;
  II[19]=   6   ;  JJ[19]=  10  ;  an[19]=   0.12324579690832e-6   ;
  II[20]=   6   ;  JJ[20]=  12  ;  an[20]=  -0.11606921130984e-5   ;
  II[21]=   6   ;  JJ[21]=  16  ;  an[21]=   0.27846367088554e-4   ;
  II[22]=   6   ;  JJ[22]=  20  ;  an[22]=  -0.59270038474176e-3   ;
  II[23]=   6   ;  JJ[23]=  22  ;  an[23]=   0.12918582991878e-2   ;
  eta = SP.h / Hn;
  T = 0.0;
  for(i=1;i<=23;i++){
          T = T + an[i]*Math.pow(SP.P+25.0,II[i])*Math.pow(eta-1.8,JJ[i]);
  }
  SP.T=T;
  
  return 1;
}


function Tps2A(SP){
  var i;
  var Sn;
  var sigma;
  var T;  
  var II = [];
  var JJ = [];
  var an = [];
  
  Sn= 2.0;
  II[ 1]=  -1.5e+0    ;  JJ[ 1]=  -24  ;  an[ 1]=  -0.39235983861984e+6  ;
  II[ 2]=  -1.5e+0    ;  JJ[ 2]=  -23  ;  an[ 2]=   0.51526573827270e+6  ;
  II[ 3]=  -1.5e+0    ;  JJ[ 3]=  -19  ;  an[ 3]=   0.40482443161048e+5  ;
  II[ 4]=  -1.5e+0    ;  JJ[ 4]=  -13  ;  an[ 4]=  -0.32193790923902e+3  ;
  II[ 5]=  -1.5e+0    ;  JJ[ 5]=  -11  ;  an[ 5]=   0.96961424218694e+2  ;
  II[ 6]=  -1.5e+0    ;  JJ[ 6]=  -10  ;  an[ 6]=  -0.22867846371773e+2  ;
  II[ 7]=  -1.25e+0   ;  JJ[ 7]=  -19  ;  an[ 7]=  -0.44942914124357e+6  ;
  II[ 8]=  -1.25e+0   ;  JJ[ 8]=  -15  ;  an[ 8]=  -0.50118336020166e+4  ;
  II[ 9]=  -1.25e+0   ;  JJ[ 9]=  -6   ;  an[ 9]=   0.35684463560015e+0  ;
  II[10]=  -1.0e+0    ;  JJ[10]=  -26  ;  an[10]=   0.44235335848190e+5  ;
  II[11]=  -1.0e+0    ;  JJ[11]=  -21  ;  an[11]=  -0.13673388811708e+5  ;
  II[12]=  -1.0e+0    ;  JJ[12]=  -17  ;  an[12]=   0.42163260207864e+6  ;
  II[13]=  -1.0e+0    ;  JJ[13]=  -16  ;  an[13]=   0.22516925837475e+5  ;
  II[14]=  -1.0e+0    ;  JJ[14]=  -9   ;  an[14]=   0.47442144865646e+3  ;
  II[15]=  -1.0e+0    ;  JJ[15]=  -8   ;  an[15]=  -0.14931130797647e+3  ;
  II[16]=  -0.75e+0   ;  JJ[16]=  -15  ;  an[16]=  -0.19781126320452e+6  ;
  II[17]=  -0.75e+0   ;  JJ[17]=  -14  ;  an[17]=  -0.23554399470760e+5  ;
  II[18]=  -0.5e+0    ;  JJ[18]=  -26  ;  an[18]=  -0.19070616302076e+5  ;
  II[19]=  -0.5e+0    ;  JJ[19]=  -13  ;  an[19]=   0.55375669883164e+5  ;
  II[20]=  -0.5e+0    ;  JJ[20]=  -9   ;  an[20]=   0.38293691437363e+4  ;
  II[21]=  -0.5e+0    ;  JJ[21]=  -7   ;  an[21]=  -0.60391860580567e+3  ;
  II[22]=  -0.25e+0   ;  JJ[22]=  -27  ;  an[22]=   0.19363102620331e+4  ;
  II[23]=  -0.25e+0   ;  JJ[23]=  -25  ;  an[23]=   0.42660643698610e+4  ;
  II[24]=  -0.25e+0   ;  JJ[24]=  -11  ;  an[24]=  -0.59780638872718e+4  ;
  II[25]=  -0.25e+0   ;  JJ[25]=  -6   ;  an[25]=  -0.70401463926862e+3  ;
  II[26]=   0.25e+0   ;  JJ[26]=  1    ;  an[26]=   0.33836784107553e+3  ;
  II[27]=   0.25e+0   ;  JJ[27]=  4    ;  an[27]=   0.20862786635187e+2  ;
  II[28]=   0.25e+0   ;  JJ[28]=  8    ;  an[28]=   0.33834172656196e-1  ;
  II[29]=   0.25e+0   ;  JJ[29]=  11   ;  an[29]=  -0.43124428414893e-4  ;
  II[30]=   0.5e+0    ;  JJ[30]=  0    ;  an[30]=   0.16653791356412e+3  ;
  II[31]=   0.5e+0    ;  JJ[31]=  1    ;  an[31]=  -0.13986292055898e+3  ;
  II[32]=   0.5e+0    ;  JJ[32]=  5    ;  an[32]=  -0.78849547999872e+0  ;
  II[33]=   0.5e+0    ;  JJ[33]=  6    ;  an[33]=   0.72132411753872e-1  ;
  II[34]=   0.5e+0    ;  JJ[34]=  10   ;  an[34]=  -0.59754839398283e-2  ;
  II[35]=   0.5e+0    ;  JJ[35]=  14   ;  an[35]=  -0.12141358953904e-4  ;
  II[36]=   0.5e+0    ;  JJ[36]=  16   ;  an[36]=   0.23227096733871e-6  ;
  II[37]=   0.75e+0   ;  JJ[37]=  0    ;  an[37]=  -0.10538463566194e+2  ;
  II[38]=   0.75e+0   ;  JJ[38]=  4    ;  an[38]=   0.20718925496502e+1  ;
  II[39]=   0.75e+0   ;  JJ[39]=  9    ;  an[39]=  -0.72193155260427e-1  ;
  II[40]=   0.75e+0   ;  JJ[40]=  17   ;  an[40]=   0.20749887081120e-6  ;
  II[41]=   1.0e+0    ;  JJ[41]=  7    ;  an[41]=  -0.18340657911379e-1  ;
  II[42]=   1.0e+0    ;  JJ[42]=  18   ;  an[42]=   0.29036272348696e-6  ;
  II[43]=   1.25e+0   ;  JJ[43]=  3    ;  an[43]=   0.21037527893619e+0  ;
  II[44]=   1.25e+0  ;  JJ[44]=  15  ;  an[44]=   0.25681239729999e-3  ;
  II[45]=   1.5e+0    ;  JJ[45]=  5    ;  an[45]=  -0.12799002933781e-1  ;
  II[46]=   1.5e+0    ;  JJ[46]=  18  ;  an[46]=  -0.82198102652018e-5  ;


  sigma = SP.s / Sn;
  T = 0.0;
  if(SP.P<=0.0){SP = null;return -1;}
  
  for(i=1;i<=46;i++){
    T = T + an[i]*pow(SP.P,II[i])*Math.pow(sigma-2.0,JJ[i]);
  }
  SP.T=T;
  
  return 1;
}

function Tps2B(SP){
  var i;
  var Sn;
  var sigma;
  var T;  
  var II = [];
  var JJ = [];
  var an = [];
  
  Sn=  0.7853;
  II[ 1]=  -6  ;  JJ[ 1]=   0  ;  an[ 1]=   0.31687665083497e+6  ;
  II[ 2]=  -6  ;  JJ[ 2]=  11  ;  an[ 2]=   0.20864175881858e+2  ;
  II[ 3]=  -5  ;  JJ[ 3]=   0  ;  an[ 3]=  -0.39859399803599e+6  ;
  II[ 4]=  -5  ;  JJ[ 4]=  11  ;  an[ 4]=  -0.21816058518877e+2  ;
  II[ 5]=  -4  ;  JJ[ 5]=   0  ;  an[ 5]=   0.22369785194242e+6  ;
  II[ 6]=  -4  ;  JJ[ 6]=   1  ;  an[ 6]=  -0.27841703445817e+4  ;
  II[ 7]=  -4  ;  JJ[ 7]=  11  ;  an[ 7]=   0.99207436071480e+1  ;
  II[ 8]=  -3  ;  JJ[ 8]=   0  ;  an[ 8]=  -0.75197512299157e+5  ;
  II[ 9]=  -3  ;  JJ[ 9]=   1  ;  an[ 9]=   0.29708605951158e+4  ;
  II[10]=  -3  ;  JJ[10]=  11  ;  an[10]=  -0.34406878548526e+1  ;
  II[11]=  -3  ;  JJ[11]=  12  ;  an[11]=   0.38815564249115e+0  ;
  II[12]=  -2  ;  JJ[12]=   0  ;  an[12]=   0.17511295085750e+5  ;
  II[13]=  -2  ;  JJ[13]=   1  ;  an[13]=  -0.14237112854449e+4  ;
  II[14]=  -2  ;  JJ[14]=   6  ;  an[14]=   0.10943803364167e+1  ;
  II[15]=  -2  ;  JJ[15]=  10  ;  an[15]=   0.89971619308495e+0  ;
  II[16]=  -1  ;  JJ[16]=   0  ;  an[16]=  -0.33759740098958e+4  ;
  II[17]=  -1  ;  JJ[17]=   1  ;  an[17]=   0.47162885818355e+3  ;
  II[18]=  -1  ;  JJ[18]=   5  ;  an[18]=  -0.19188241993679e+1  ;
  II[19]=  -1  ;  JJ[19]=   8  ;  an[19]=   0.41078580492196e+0  ;
  II[20]=  -1  ;  JJ[20]=   9  ;  an[20]=  -0.33465378172097e+0  ;
  II[21]=   0  ;  JJ[21]=   0  ;  an[21]=   0.13870034777505e+4  ;
  II[22]=   0  ;  JJ[22]=   1  ;  an[22]=  -0.40663326195838e+3  ;
  II[23]=   0  ;  JJ[23]=   2  ;  an[23]=   0.41727347159610e+2  ;
  II[24]=   0  ;  JJ[24]=   4  ;  an[24]=   0.21932549434532e+1  ;
  II[25]=   0  ;  JJ[25]=   5  ;  an[25]=  -0.10320050009077e+1  ;
  II[26]=   0  ;  JJ[26]=   6  ;  an[26]=   0.35882943516703e+0  ;
  II[27]=   0  ;  JJ[27]=   9  ;  an[27]=   0.52511453726066e-2  ;
  II[28]=   1  ;  JJ[28]=   0  ;  an[28]=   0.12838916450705e+2  ;
  II[29]=   1  ;  JJ[29]=   1  ;  an[29]=  -0.28642437219381e+1  ;
  II[30]=   1  ;  JJ[30]=   2  ;  an[30]=   0.56912683664855e+0  ;
  II[31]=   1  ;  JJ[31]=   3  ;  an[31]=  -0.99962954584931e-1  ;
  II[32]=   1  ;  JJ[32]=   7  ;  an[32]=  -0.32632037778459e-2  ;
  II[33]=   1  ;  JJ[33]=   8  ;  an[33]=   0.23320922576723e-3  ;
  II[34]=   2  ;  JJ[34]=   0  ;  an[34]=  -0.15334809857450e+0  ;
  II[35]=   2  ;  JJ[35]=   1  ;  an[35]=   0.29072288239902e-1  ;
  II[36]=   2  ;  JJ[36]=   5  ;  an[36]=   0.37534702741167e-3  ;
  II[37]=   3  ;  JJ[37]=   0  ;  an[37]=   0.17296691702411e-2  ;
  II[38]=   3  ;  JJ[38]=   1  ;  an[38]=  -0.38556050844504e-3  ;
  II[38]=   3  ;  JJ[38]=   1  ;  an[38]=  -0.38556050844504e-3  ;
  II[39]=   3  ;  JJ[39]=   3  ;  an[39]=  -0.35017712292608e-4  ;
  II[40]=   4  ;  JJ[40]=   0  ;  an[40]=  -0.14566393631492e-4  ;
  II[41]=   4  ;  JJ[41]=   1  ;  an[41]=   0.56420857267269e-5  ;
  II[42]=   5  ;  JJ[42]=   0  ;  an[42]=   0.41286150074605e-7  ;
  II[43]=   5  ;  JJ[43]=   1  ;  an[43]=  -0.20684671118824e-7  ;
  II[44]=   5  ;  JJ[44]=   2  ;  an[44]=   0.16409393674725e-8  ;
  
  sigma = SP.s / Sn;
  T = 0.0;
  for(i=1;i<=44;i++){
    T = T + an[i]*Math.pow(SP.P,II[i])*Math.pow(10.0-sigma,JJ[i]);
  }
  SP.T=T;
  
  return 1;
}

function Tps2C(SP){
  var i;
  var Sn;
  var sigma;
  var T;  
  var II = [];
  var JJ = [];
  var an = [];
  
  Sn=  2.9251;
  II[ 1]=  -2  ;  JJ[ 1]=  0  ;  an[ 1]=   0.90968501005365e+3  ;
  II[ 2]=  -2  ;  JJ[ 2]=  1  ;  an[ 2]=   0.24045667088420e+4  ;
  II[ 3]=  -1  ;  JJ[ 3]=  0  ;  an[ 3]=  -0.59162326387130e+3  ;
  II[ 4]=   0  ;  JJ[ 4]=  0  ;  an[ 4]=   0.54145404128074e+3  ;
  II[ 5]=   0  ;  JJ[ 5]=  1  ;  an[ 5]=  -0.27098308411192e+3  ;
  II[ 6]=   0  ;  JJ[ 6]=  2  ;  an[ 6]=   0.97976525097926e+3  ;
  II[ 7]=   0  ;  JJ[ 7]=  3  ;  an[ 7]=  -0.46966772959435e+3  ;
  II[ 8]=   1  ;  JJ[ 8]=  0  ;  an[ 8]=   0.14399274604723e+2  ;
  II[ 9]=   1  ;  JJ[ 9]=  1  ;  an[ 9]=  -0.19104204230429e+2  ;
  II[10]=   1  ;  JJ[10]=  3  ;  an[10]=   0.53299167111971e+1  ;
  II[11]=   1  ;  JJ[11]=  4  ;  an[11]=  -0.21252975375934e+2  ;
  II[12]=   2  ;  JJ[12]=  0  ;  an[12]=  -0.31147334413760e+0  ;
  II[13]=   2  ;  JJ[13]=  1  ;  an[13]=   0.60334840894623e+0  ;
  II[14]=   2  ;  JJ[14]=  2  ;  an[14]=  -0.42764839702509e-1  ;
  II[15]=   3  ;  JJ[15]=  0  ;  an[15]=   0.58185597255259e-2  ;
  II[16]=   3  ;  JJ[16]=  1  ;  an[16]=  -0.14597008284753e-1  ;
  II[17]=   3  ;  JJ[17]=  5  ;  an[17]=   0.56631175631027e-2  ;
  II[18]=   4  ;  JJ[18]=  0  ;  an[18]=  -0.76155864584577e-4  ;
  II[19]=   4  ;  JJ[19]=  1  ;  an[19]=   0.22440342919332e-3  ;
  II[20]=   4  ;  JJ[20]=  4  ;  an[20]=  -0.12561095013413e-4  ;
  II[21]=   5  ;  JJ[21]=  0  ;  an[21]=   0.63323132660934e-6  ;
  II[22]=   5  ;  JJ[22]=  1  ;  an[22]=  -0.20541989675375e-5  ;
  II[23]=   5  ;  JJ[23]=  2  ;  an[23]=   0.36405370390082e-7  ;
  II[24]=   6  ;  JJ[24]=  0  ;  an[24]=  -0.29759897789215e-8  ;
  II[25]=   6  ;  JJ[25]=  1  ;  an[25]=   0.10136618529763e-7  ;
  II[26]=   7  ;  JJ[26]=  0  ;  an[26]=   0.59925719692351e-11 ;
  II[27]=   7  ;  JJ[27]=  1  ;  an[27]=  -0.20677870105164e-10 ;
  II[28]=   7  ;  JJ[28]=  3  ;  an[28]=  -0.20874278181886e-10 ;
  II[29]=   7  ;  JJ[29]=  4  ;  an[29]=   0.10162166825089e-9  ;
  II[30]=   7  ;  JJ[30]=  5  ;  an[30]=  -0.16429828281347e-9  ;
  
  sigma = SP.s / Sn;
  T = 0.0;
  for(i=1;i<=30;i++){
    T = T + an[i]*Math.pow(SP.P,II[i])*Math.pow(2.0-sigma,JJ[i]);
  }
  SP.T=T;
  
  return 1;
}
